tests/draft2019-09/recursiveRef.json

```diff
@@ -0,0 +1,273 @@
+                        }
+                    ]
+                }
+            },
+            "anyOf": [
+                { "type": "integer" },
+                { "$ref": "#/$defs/myobject" }
+            ]
+        },
+        "tests": [
+            {
+                "description": "integer matches at the outer level",
+                "data": 1,
+                "valid": true
+            },
+            {
+                "description": "single level match",
+                "data": { "foo": "hi" },
+                "valid": true
+            },
+            {
+                "description": "integer does not match as a property value",
+                "data": { "foo": 1 },
+                "valid": false
+            },
+            {
+                "description": "two levels, properties match with inner definition",
+                "data": { "foo": { "bar": "hi" } },
+                "valid": true
+            },
+            {
+                "description": "two levels, integer does not match as a property value",
+                "data": { "foo": { "bar": 1 } },
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "$recursiveRef with no $recursiveAnchor works like $ref",
+        "schema": {
+            "$id": "http://localhost:4242/recursiveRef5/schema.json",
+            "$defs": {
+                "myobject": {
+                    "$id": "myobject.json",
+                    "$recursiveAnchor": false,
+                    "anyOf": [
+                        { "type": "string" },
+                        {
+                            "type": "object",
+                            "additionalProperties": { "$recursiveRef": "#" }
+                        }
+                    ]
+                }
+            },
+            "anyOf": [
+                { "type": "integer" },
+                { "$ref": "#/$defs/myobject" }
+            ]
+        },
+        "tests": [
+            {
+                "description": "integer matches at the outer level",
+                "data": 1,
+                "valid": true
+            },
+            {
+                "description": "single level match",
+                "data": { "foo": "hi" },
+                "valid": true
+            },
+            {
+                "description": "integer does not match as a property value",
+                "data": { "foo": 1 },
+                "valid": false
+            },
+            {
+                "description": "two levels, properties match with inner definition",
+                "data": { "foo": { "bar": "hi" } },
+                "valid": true
+            },
+            {
+                "description": "two levels, integer does not match as a property value",
+                "data": { "foo": { "bar": 1 } },
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "$recursiveRef with no $recursiveAnchor in the initial target schema resource",
```

Comment:
This test is wrong as discussed in previous comments. Let's fix or remove this test, then I would love to see this merged!

Reply:
@jdesrosiers Could you help me out by pointing to which tests are incorrect?  I do not see any that are wrong. (revisions for $dynamicRef do not apply to this version of the spec.)

specifically, regarding recursiveRef6, there is no $recursiveAnchor: true at the initial $recursiveRef target, so it behaves like a regular $ref, and we only recurse to the inner subschema, and never the outer one.

Reply:
> we need to have $recursiveAnchor: true in both the base and inner subschemas in order to recurse to the base

That's not correct. That's what I was pointing out in https://github.com/json-schema-org/JSON-Schema-Test-Suite/pull/391#issuecomment-683924227 (and Henry confirmed was the correct interpretation). `$recursiveRef` should use the first `$recursiveAnchor` in the heirarchy that was set. `$recursiveAnchor` is not required to be in every schema `$recursiveRef` appears in.

So, the second and third tests are wrong as well as the `$comment`.

(If this needs more discussion, I suggest removing these tests for now and getting the rest merged. We can add these back separately when everyone is in agreement.)

Reply:
Okay, so we're talking about the test schemas with ids `http://localhost:4242/recursiveRef2/schema.json` and `http://localhost:4242/recursiveRef3/schema.json`? I need to be sure if I'm going to merge "the rest".

> $recursiveAnchor is not required to be in every schema $recursiveRef appears in.

I don't see any tests that assert that in comments, nor assume that in their schemas. While all the $recursiveRef targets in these tests are '#', that resolves to a different position than the schema $recursiveRef appears in (because of the use of allOf). What *is* required is that the initial $recursiveRef target have a $recursiveAnchor: true in order to consider earlier/outer $recursiveAnchors, and all the tests are properly relying on that.

Please could you be specific about what schemas and subtests are not behaving properly? This is important not just to get the tests right, but to fix all the implementations that are assuming the tests are right and presently validate against these tests.

Reply:
No, we're talking about the group of tests that this comment is on. The one with `"description": "$recursiveRef with no $recursiveAnchor in the initial target schema resource",`. The one with `http://localhost:4242/recursiveRef6/inner.json`. 

I'll comment on the specific lines in question to make sure it can't be missed, but I'd suggest leaving out this whole group of tests for now as the premise of the test is wrong.